File: algorithms/2_medium/C/02116_check_if_a_parentheses_string_can_be_valid/solution.c
typedef ptrdiff_t isize;

inline static bool is_ok_forward(const char *restrict s, const char *restrict l, const isize len) {
    isize balance = 0;

    for (isize i = 0; i < len; i += 1) {
        if (l[i] == '1' && s[i] == ')') {
            balance -= 1;
        } else {
            balance += 1;
        }

        if (balance < 0) {
            return false;
        }
    }

    return true;
}

inline static bool is_ok_backward(const char *restrict s, const char *restrict l, const isize len) {
    isize balance = 0;

    for (isize i = len - 1; i >= 0; i -= 1) {
        if (l[i] == '1' && s[i] == '(') {
            balance -= 1;
        } else {
            balance += 1;
        }

        if (balance < 0) {
            return false;
        }
    }

    return true;
}

bool canBeValid(const char *restrict s, const char *restrict locked) {
    const isize len = (isize)strlen(s);

    if ((len & 1) == 1) { return false; }

    return is_ok_forward(s, locked, len) && is_ok_backward(s, locked, len);
}

File: algorithms/2_medium/C/02116_check_if_a_parentheses_string_can_be_valid/solution.rs
impl Solution {
    #[inline(always)]
    fn is_ok_forward(s: &[u8], l: &[u8]) -> bool {
        let len = s.len();
        let mut balance = 0_isize;

        for i in 0..len {
            match (l[i], s[i]) {
                (b'1', b')') => balance -= 1,
                _ => balance += 1,
            }

            if balance < 0 {
                return false;
            }
        }

        true
    }

    #[inline(always)]
    fn is_ok_backward(s: &[u8], l: &[u8]) -> bool {
        let len = s.len();
        let mut balance = 0_isize;

        for i in (0..len).rev() {
            match (l[i], s[i]) {
                (b'1', b'(') => balance -= 1,
                _ => balance += 1,
            }

            if balance < 0 {
                return false;
            } 
        }

        true
    }

    pub fn can_be_valid(s: String, locked: String) -> bool {
        debug_assert!(s.len() == locked.len() && !s.is_empty());

        if (s.len() & 1) == 1 { return false; }

        Self::is_ok_forward(s.as_bytes(), locked.as_bytes()) &&
        Self::is_ok_backward(s.as_bytes(), locked.as_bytes())
    }
}
